test/test_distributions.py

```diff
@@ -4092,5 +4115,161 @@ def tearDown(self):
         super(TestCase, self).tearDown()
         Distribution.set_default_validate_args(False)
 
+
+class TestJit(TestCase):
+    def _examples(self):
+        for Dist, params in EXAMPLES:
+            for param in params:
+                print('testing {}'.format(Dist.__name__))
+                keys = param.keys()
+                values = tuple(param[key] for key in keys)
+                if not all(isinstance(x, torch.Tensor) for x in values):
+                    continue
+                sample = Dist(**param).sample()
+                yield Dist, keys, values, sample
+
+    def test_sample(self):
+        for Dist, keys, values, sample in self._examples():
+
+            def f(*values):
+                param = dict(zip(keys, values))
+                dist = Dist(**param)
+                return dist.sample()
+
+            traced_f = torch.jit.trace(f, values, check_trace=False)
+
+            # FIXME Schema not found for node
+            xfail = [
+                Cauchy,  # aten::cauchy(Double(2,1), float, float, Generator)
+                HalfCauchy,  # aten::cauchy(Double(2, 1), float, float, Generator)
+            ]
+            if Dist in xfail:
+                print('        xfail reproducibility')
+                continue
+
+            with torch.random.fork_rng():
+                sample_0 = traced_f(*values)
+            sample_1 = traced_f(*values)
+            self.assertEqual(sample_0, sample_1)
+
+            # FIXME no nondeterministic nodes found in trace
+            xfail = [Beta, Dirichlet]
+            if Dist in xfail:
+                print('        xfail nondeterministic nodes')
+                continue
+            self.assertTrue(any(n.isNondeterministic() for n in traced_f.graph.nodes()))
+
+    def test_rsample(self):
+        for Dist, keys, values, sample in self._examples():
+            if not Dist.has_rsample:
+                continue
+
+            def f(*values):
+                param = dict(zip(keys, values))
+                dist = Dist(**param)
+                return dist.rsample()
+
+            traced_f = torch.jit.trace(f, values, check_trace=False)
+
+            # FIXME Schema not found for node
+            xfail = [
+                Cauchy,  # aten::cauchy(Double(2,1), float, float, Generator)
+                HalfCauchy,  # aten::cauchy(Double(2, 1), float, float, Generator)
+            ]
+            if Dist in xfail:
+                print('        xfail reproducibility')
+                continue
+
+            with torch.random.fork_rng():
+                sample_0 = traced_f(*values)
+            sample_1 = traced_f(*values)
```

Comment:
This assert is not really checking anything, which explains why this condition passes even though there are no random nodes in the graph. I meant that we should compare `f` (under forked rng) with `traced_f`

Reply:
We can't generally compare `f` to `traced_f` because some samplers use different strategies based on `torch._C._get_tracing_State()`. I guess I'll just add exceptions for those cases.

Reply:
Yeah that would be better. Comparing `traced_f` to `traced_f` doesn't make much sense.